### PR TITLE
Add basic jupyter notebook support

### PR DESCRIPTION
**Pull request by konstin**
This feature is disabled by default for now behind the `jupyter_notebook` feature.

Missing:
 * do something about autofixing
 * writing back notebooks while making sure we don't change them in the process (round trip test)

### CODE REVIEW

File: crates/ruff/src/jupyter/schema.rs [hunk 1]
```diff
@@ -0,0 +1,253 @@
+//! The json schema of a jupyter notebook, entrypoint is [JupyterNotebook]
+//!
+//! Generated by https://app.quicktype.io/ from
+//! <https://github.com/jupyter/nbformat/blob/16b53251aabf472ad9406ddb1f78b0421c014eeb/nbformat/v4/nbformat.v4.schema.json>
+//! Jupyter Notebook v4.5 JSON schema.
+//!
+//! Some manual edits
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/schema.rs:7]: What manual edits? Could we separate the manual edits into their own commit, so that even if the PR is squashed, we have access to the history here?

File: crates/ruff/src/resolver.rs [hunk 2]
```diff
@@ -192,15 +193,15 @@ fn match_exclusion(file_path: &str, file_basename: &str, exclusion: &globset::Gl
     exclusion.is_match(file_path) || exclusion.is_match(file_basename)
 }
 
-/// Return `true` if the [`DirEntry`] appears to be that of a Python file.
+/// Return `true` if the [`DirEntry`] appears to be that of a Python file or jupyter notebook.
 pub fn is_python_entry(entry: &DirEntry) -> bool {
-    is_python_file(entry.path())
+    (is_python_file(entry.path()) || is_jupyter_notebook(entry.path()))
```
Comment (charliermarsh) [inline crates/ruff/src/resolver.rs:198]: Cool - so this will go behind a feature flag as discussed synchronously?

File: crates/ruff_cli/Cargo.toml [hunk 3]
```diff
@@ -59,6 +60,7 @@ walkdir = { version = "2.3.2" }
 
 [dev-dependencies]
 assert_cmd = { version = "2.0.8" }
+indoc = "2.0.1"
```
Comment (charliermarsh) [inline crates/ruff_cli/Cargo.toml:63]: Nit: we tend to use the expanded version everywhere (I don't know why, but it is the convention).

File: crates/ruff_cli/src/diagnostics.rs [hunk 4]
```diff
@@ -83,7 +91,25 @@ pub fn lint_path(
...
     // Read the file from disk.
-    let contents = std::fs::read_to_string(path)?;
+    let (contents, jupyter_index) = if is_jupyter_notebook(&path) {
+        // See also the black implementation
+        // <https://github.com/psf/black/blob/69ca0a4c7a365c5f5eea519a90980bab72cab764/src/black/__init__.py#L1017-L1046>
+        let notebook: JupyterNotebook = if let Some(notebook) = read_jupyter_notebook(path)? {
+            notebook
+        } else {
+            // Not a python notebook
+            return Ok(Diagnostics {
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:101]: Should this be returning some kind of error diagnostic?

File: crates/ruff_cli/src/printer.rs [hunk 5]
```diff
@@ -691,13 +731,29 @@ fn print_grouped_message<T: Write>(
     autofix_level: fix::FixMode,
     row_length: usize,
     column_length: usize,
+    jupyter_index: Option<&JupyterIndex>,
 ) -> Result<()> {
+    // Check if we're working on a jupyter notebook and translate positions with cell accordingly
+    let pos_label = if let Some(jupyter_index) = jupyter_index {
+        format!(
+            "cell {cell}{sep}{row}{sep}{col}",
```
Comment (charliermarsh) [inline crates/ruff_cli/src/printer.rs:739]: How does nbqa format / convey these?

File: crates/ruff/src/jupyter/jupyter.rs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        );
+        return Ok(None);
+    }
+
+    Ok(Some(notebook))
+}
+
+/// Concatenates all cells into a single virtual file and builds an index that maps the content
+/// to notebook cell locations
+pub fn concat_notebook(notebook: JupyterNotebook) -> (String, JupyterIndex) {
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/jupyter.rs:97]: Can we add some tests for this? It looks unit-testable.

File: crates/ruff/src/jupyter/jupyter.rs [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+                    return Err(err).context(format!(
+                        "Expected a jupyter notebook (.{} extension), \
+                        which must be internally stored as json, \
+                        but found a python source file",
+                        JUPYTER_NOTEBOOK_EXT
+                    ));
+                }
+
+                return Err(err).context(format!(
+                    "A jupyter notebook (.{}) must internally be json, \
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/jupyter.rs:57]: Nit: IMO we should capitalize as `Jupyter Notebook`, `JSON`, and `Python` for any user-facing text. (For "correctness", but also for consistency with other user-facing text.)

File: crates/ruff/src/jupyter/jupyter.rs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                // We could try to read the schema version here but if this fails it's
+                // a bug anyway
+                return Err(err).context(format!(
+                    "This file does not match the schema expected of jupyter notebooks"
+                ));
+            }
+        },
+    };
+
+    if notebook.nbformat != 4 {
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/jupyter.rs:72]: Mind adding a comment here to explain what format 4 is vs. other values that this might be?

File: crates/ruff_cli/src/diagnostics.rs [hunk 9]
```diff
@@ -1,35 +1,42 @@
...
 use crate::cache;
+use ruff::jupyter::is_jupyter_notebook;
 
 #[derive(Debug, Default)]
 pub struct Diagnostics {
     pub messages: Vec<Message>,
     pub fixed: FxHashMap<String, FixTable>,
+    /// Jupyter notebook indexing table for each input file that is a jupyter notebook
+    /// so we can rewrite the diagnostics in the end
+    pub jupyter_index: HashMap<String, JupyterIndex>,
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:31]: Rather than passing around the index... I'm wondering if we should instead change `Message` to have a more flexible location type, and store the location data on `Message` directly. This isn't a fully-formed idea, but it seems non-ideal that clients of `Diagnostics` need to do this translation, worry about what's a notebook, what isn't, etc. Could `Message` instead have a `Location` that's an enum or similar, and can be (e.g.) a Python file location or a Jupyter notebook location? What would be the tradeoffs?

In other words, this feels like exposing an implementation detail, since clients only care about the location, and not that translation step. What do you think?

File: crates/ruff_cli/src/diagnostics.rs [hunk 10]
```diff
@@ -1,35 +1,42 @@
...
 }
 
 impl Diagnostics {
     pub fn new(messages: Vec<Message>) -> Self {
         Self {
             messages,
             fixed: FxHashMap::default(),
+            jupyter_index: HashMap::new(),
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:39]: Any reason not to use `FxHashMap`?

File: crates/ruff_cli/src/commands/run.rs [hunk 11]
```diff
@@ -137,3 +144,85 @@ pub fn run(
...
+
+        let printer = Printer::new(
+            SerializationFormat::Text,
+            LogLevel::Default,
+            FixMode::None,
+            Flags::SHOW_VIOLATIONS,
+        );
+        let mut writer: Vec<u8> = Vec::new();
+        printer.write_once(&diagnostics, &mut writer)?;
+        // TODO: Set jupyter notebooks as none-fixable for now
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/run.rs:213]: Do you mind using the `TODO(konsti)`-style TODOs? I like to have authors indicated in the TODOs, not as a sign that they're responsible for fixing them, but as a guidepost for who wrote it and has context on it.

File: crates/ruff/src/jupyter/jupyter.rs [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+                    .last()
+                    .map_or(true, |last| last.is_err())
+                {
+                    return Err(err).context(format!(
+                        "Expected a jupyter notebook (.{} extension), \
+                        which must be internally stored as json, \
+                        but found a python source file",
+                        JUPYTER_NOTEBOOK_EXT
+                    ));
+                }
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/jupyter.rs:54]: Oh interesting, so this is checking that, in the event that we failed to parse, did the user accidentally save a Python file as a `.ipynb` file?

File: crates/ruff/src/jupyter/jupyter.rs [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+pub fn concat_notebook(notebook: JupyterNotebook) -> (String, JupyterIndex) {
+    let mut jupyter_index = JupyterIndex {
+        // Enter a line number (1-based), get back the cell (1-based)
+        // 0 index is just padding
+        row_to_cell: vec![0],
+        // Enter a line number (1-based), get back the row number in the cell (1-based)
+        // 0 index is just padding
+        row_to_row_in_cell: vec![0],
+    };
+    let mut contents = Vec::new();
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/jupyter.rs:106]: Is there any way to hint the size here? (Maybe not!)

File: crates/ruff/src/jupyter/jupyter.rs [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    Ok(Some(notebook))
+}
+
+/// Concatenates all cells into a single virtual file and builds an index that maps the content
+/// to notebook cell locations
+pub fn concat_notebook(notebook: JupyterNotebook) -> (String, JupyterIndex) {
+    let mut jupyter_index = JupyterIndex {
+        // Enter a line number (1-based), get back the cell (1-based)
+        // 0 index is just padding
+        row_to_cell: vec![0],
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/jupyter.rs:101]: What's the motivation for 1-based-to-1-based, rather than 0-based to 0-based? Seems like it could potentially cause confusion to have this 0 padding.

File: crates/ruff_cli/src/diagnostics.rs [hunk 15]
```diff
@@ -83,7 +91,25 @@ pub fn lint_path(
     };
 
     // Read the file from disk.
-    let contents = std::fs::read_to_string(path)?;
+    let (contents, jupyter_index) = if is_jupyter_notebook(&path) {
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:94]: Idly wondering if we should tag files by dialect in the `resolver`, and pass around tagged files, rather than checking extensions on `Path` everywhere. Same goes for the `.pyi` files. (Maybe not actionable, just a rough thought.)

**Review by charliermarsh — COMMENTED** [review]
Sweet! Impressive that you were able to tease apart the existing codepaths to get this far. I know we chatted about a few changes here earlier (feature flag, maybe using diagnostics in lieu of some of the current error handling), but thought I'd drop some comments on current state :)

File: crates/ruff_cli/Cargo.toml (hunk 3 above)
Comment (konstin) [reply crates/ruff_cli/Cargo.toml:63]: tbh i'm a big fan of indoc everywhere, but yes consistency

File: crates/ruff_cli/src/diagnostics.rs (hunk 4 above)
Comment (konstin) [reply crates/ruff_cli/src/diagnostics.rs:101]: updated the comment, this could e.g. be an R notebook which we want to just skip

File: crates/ruff_cli/src/printer.rs (hunk 5 above)
Comment (konstin) [reply crates/ruff_cli/src/printer.rs:739]:
```
$ nbqa ruff /home/konsti/ruff/crates/ruff/resources/test/fixtures/jupyter/
crates/ruff/resources/test/fixtures/jupyter/valid.ipynb:cell_1:2:5: F841 [*] Local variable `x` is assigned to but never used
Found 1 error.
[*] 1 potentially fixable with the --fix option.

nbQA failed to process /home/konsti/ruff/crates/ruff/resources/test/fixtures/jupyter/not_json.ipynb with exception "JSONDecodeError('Expecting value: line 1 column 1 (char 0)')"
nbQA failed to process /home/konsti/ruff/crates/ruff/resources/test/fixtures/jupyter/invalid_extension.ipynb with exception "JSONDecodeError('Expecting value: line 1 column 1 (char 0)')"
nbQA failed to process /home/konsti/ruff/crates/ruff/resources/test/fixtures/jupyter/wrong_schema.ipynb with exception "KeyError('cells')"

If you believe the notebook(s) to be valid, please report a bug at https://github.com/nbQA-dev/nbQA/issues 
```

Underscore instead of space but otherwise same format

File: crates/ruff/src/jupyter/jupyter.rs (hunk 8 above)
Comment (konstin) [reply crates/ruff/src/jupyter/jupyter.rs:72]: tbh i've only seen v4 in the wild

File: crates/ruff_cli/src/commands/run.rs (hunk 11 above)
Comment (JonathanPlasse) [reply crates/ruff_cli/src/commands/run.rs:213]: What about git blame? Is it in case of refactoring?

File: crates/ruff_cli/src/diagnostics.rs (hunk 9 above)
Comment (konstin) [reply crates/ruff_cli/src/diagnostics.rs:31]: i'm not completely sure either; my current thought is that most code can just assume to see one consistent python file and not bother with everything else and a special case such as jupyter notebooks gets patched in the appropriate places.

making `location` and `end_location` an enum of `Location(Location)` and `JupyterLocation { cell: u32, row: u32, column: u32}` would definitely also be possible

File: crates/ruff/src/jupyter/jupyter.rs (hunk 12 above)
Comment (konstin) [reply crates/ruff/src/jupyter/jupyter.rs:54]: yes

File: crates/ruff/src/jupyter/jupyter.rs (hunk 13 above)
Comment (konstin) [reply crates/ruff/src/jupyter/jupyter.rs:106]: i've also checked that `.join("\n")` does size hints internally

File: crates/ruff/src/jupyter/jupyter.rs (hunk 14 above)
Comment (konstin) [reply crates/ruff/src/jupyter/jupyter.rs:101]: mainly that everywhere else uses 1-based indices

File: crates/ruff_cli/src/diagnostics.rs (hunk 15 above)
Comment (konstin) [reply crates/ruff_cli/src/diagnostics.rs:94]: maybe a  `enum FileInfo { Py, PyI, Jupyter { index: JupyterIndex }}`, we could also put that in a hash map `Path` (or string) -> `FileInfo` that we read when collecting the results

File: crates/ruff/src/jupyter/schema.rs (hunk 1 above)
Comment (konstin) [reply crates/ruff/src/jupyter/schema.rs:7]: i've added comments on what i changed

File: crates/ruff/src/resolver.rs (hunk 2 above)
Comment (konstin) [reply crates/ruff/src/resolver.rs:198]: yes, in `is_jupyter_notebook`

File: crates/ruff_cli/src/printer.rs
Comment (konstin) [inline crates/ruff_cli/src/printer.rs:1]: this is a mess unfortunately

**Comment by konstin** [comment]
Thank you for the detailed review!

File: crates/ruff/src/jupyter/schema.rs [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+/// Notebook raw nbconvert cell.
+///
+/// Notebook markdown cell.
+///
+/// Notebook code cell.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Cell {
+    pub attachments: Option<HashMap<String, HashMap<String, SourceValue>>>,
```
Comment (MichaReiser) [inline crates/ruff/src/jupyter/schema.rs:44]: Is the difference between an empty `HashMap` (and an empty `Vec` for `outputs`) important or can we change the types to `HashMap<String<HashMap<String, SourceValue>>` and use `skip_if=HashMapp::is_empty`?

The same applies for other `Option<Collection>` usages in this file

File: crates/ruff/src/jupyter/notebook.rs [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+        );
+        return Ok(None);
+    }
+
+    Ok(Some(notebook))
+}
+
+/// Concatenates all cells into a single virtual file and builds an index that maps the content
+/// to notebook cell locations
+pub fn concat_notebook(notebook: &JupyterNotebook) -> (String, JupyterIndex) {
```
Comment (MichaReiser) [inline crates/ruff/src/jupyter/notebook.rs:147]: Nit: I expected this function to concat two notebooks. What do you think of adding a `index()` function on `JupyterNotebook`?

File: crates/ruff_cli/src/diagnostics.rs [hunk 18]
```diff
@@ -85,7 +92,39 @@ pub fn lint_path(
...
+                return Ok(Diagnostics {
+                    messages: vec![Message::from_diagnostic(
+                        *diagnostic,
+                        path.to_string_lossy().to_string(),
+                        None,
+                        1,
+                    )],
+                    fixed: FxHashMap::default(),
+                    jupyter_index: FxHashMap::default(),
+                });
```
Comment (MichaReiser) [inline crates/ruff_cli/src/diagnostics.rs:119]: You can simplify this if `Diagnostics` implements `Default`
```suggestion
                return Ok(Diagnostics {
                	..Diagnostics::default(),
                	messages: vec![Message::from_diagnostic(
                        *diagnostic,
                        path.to_string_lossy().to_string(),
                        None,
                        1,
                    )],
                });
```

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_cli/src/printer.rs
Comment (MichaReiser) [reply crates/ruff_cli/src/printer.rs:1]: Yeah, I think we should prioritize reworking our Diagnostics representations. We're hitting rough edges in many places.

File: crates/ruff/src/jupyter/notebook.rs [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+        // 0 index is just padding
+        row_to_row_in_cell: vec![0],
+    };
+    let size_hint = notebook
+        .cells
+        .iter()
+        .filter(|cell| cell.cell_type == CellType::Code)
+        .count();
+
+    let mut contents = Vec::with_capacity(size_hint);
```
Comment (MichaReiser) [inline crates/ruff/src/jupyter/notebook.rs:162]: Did you profile if setting the `size_hint` outperforms iterating over the code cells?

**Review by MichaReiser — APPROVED** [review]
Excellent work and thanks for taking the time to document the data structures. 

Could we add one or two jupyter notebooks to our benchmark so that we have good numbers from the beginning?

File: crates/ruff/src/jupyter/schema.rs (hunk 16 above)
Comment (konstin) [reply crates/ruff/src/jupyter/schema.rs:44]: afaik it saying the field to be missing vs. the field is present but empty. it believe it makes a difference for round-trip serialization and json schema compliance

File: crates/ruff/src/jupyter/notebook.rs [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+pub const JUPYTER_NOTEBOOK_EXT: &str = "ipynb";
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based [Location]
+/// to jupyter notebook cell/row/column.
+#[derive(Debug, Eq, PartialEq)]
+pub struct JupyterIndex {
+    /// Enter a row (1-based), get back the cell (1-based)
+    pub row_to_cell: Vec<usize>,
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff/src/jupyter/notebook.rs:26]: is there are a smart way to convert usize -> u32? because clippy complains

> casting `usize` to `u32` may truncate the value on targets with 64-bit wide pointers

File: crates/ruff/src/jupyter/notebook.rs (hunk 17 above)
Comment (konstin) [reply crates/ruff/src/jupyter/notebook.rs]: good point, i renamed it `index_notebook`

File: crates/ruff/src/jupyter/notebook.rs [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+                jupyter_index
+                    .row_to_cell
+                    .extend(iter::repeat(pos + 1).take(string_array.len()));
+                jupyter_index
+                    .row_to_row_in_cell
+                    .extend(1..=string_array.len());
+                // lines already end in a newline character
+                string_array.join("")
+            }
+        };
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff/src/jupyter/notebook.rs]: the contents string itself we have to build anew and the index doesn't have any string contents. i'd have hoped that the current code is already compiler friendly enough that the content string building is fast enough not to show up in any flamegraphs

File: crates/ruff/src/jupyter/notebook.rs (hunk 19 above)
Comment (konstin) [reply crates/ruff/src/jupyter/notebook.rs:151]: no i did that based on charlie's comment above; i currently also don't have any large enough jupyter notebook repo to make relevant tests

File: crates/ruff/src/jupyter/notebook.rs [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+        contents.push(cell_contents);
+    }
+    // The last line doesn't end in a newline character
+    (contents.join("\n"), jupyter_index)
+}
+
+/// Write back with an indent of 1, just like black
+pub fn write_jupyter_notebook(path: &Path, notebook: &JupyterNotebook) -> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff/src/jupyter/notebook.rs:188]: didn't think about that, that is much nicer, thanks

**Comment by konstin** [comment]
> Could we add one or two jupyter notebooks to our benchmark so that we have good numbers from the beginning?

i'll add benchmarks in the next round, this is already large

**Comment by konstin** [comment]
Thank to you too for the detailed review!

File: crates/ruff_cli/Cargo.toml [hunk 23]
```diff
@@ -60,6 +61,9 @@ walkdir = { version = "2.3.2" }
 assert_cmd = { version = "2.0.8" }
 ureq = { version = "2.6.2", features = [] }
 
+[features]
+jupyter_notebook = ["ruff/jupyter_notebook"]
```
Comment (charliermarsh) [inline crates/ruff_cli/Cargo.toml:65]: (We need to do this for `logical_lines` too I think, not part of this PR, just a note to self...)

File: crates/ruff_cli/src/diagnostics.rs [hunk 24]
```diff
@@ -165,9 +214,24 @@ pub fn lint_path(
         }
     }
 
+    let jupyter_index = match jupyter_index {
+        None => FxHashMap::default(),
+        Some(jupyter_index) => {
+            let mut index = FxHashMap::default();
+            index.insert(
+                path.to_str()
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:222]: Should we use `relativize_path` here for consistency with `fixed`, below?

File: crates/ruff_cli/src/printer.rs (hunk 5 above)
Comment (charliermarsh) [reply crates/ruff_cli/src/printer.rs:790]: Should we use an underscore? Wdyt? It might play better with regex parsers that rely on the structure of this output.

File: crates/ruff_cli/src/printer.rs
Comment (charliermarsh) [reply crates/ruff_cli/src/printer.rs:1]: I'd def support a PR that improves the Diagnostic representation with this generalization in mind.

File: crates/ruff_cli/src/diagnostics.rs [hunk 25]
```diff
@@ -48,9 +53,46 @@ impl AddAssign for Diagnostics {
...
+        Ok(notebook) => {
+            if !notebook
+                .metadata
+                .language_info
+                .as_ref()
+                .map_or(true, |language| language.name == "python")
+            {
+                // Not a python notebook, this could e.g. be an R notebook which we want to just skip
+                debug!(
+                    "Skipping {} because it's not a python notebook",
```
Comment (charliermarsh) [inline crates/ruff_cli/src/diagnostics.rs:72]: Nit: `Python` for user-facing messages.

File: crates/ruff_cli/src/commands/run.rs [hunk 26]
```diff
@@ -83,8 +83,15 @@ pub fn run(
...
                     let settings = resolver.resolve_all(path, pyproject_strategy);
-                    lint_path(path, package, settings, cache, noqa, autofix)
-                        .map_err(|e| (Some(path.to_owned()), e.to_string()))
+                    lint_path(path, package, settings, cache, noqa, autofix).map_err(|e| {
+                        (Some(path.to_owned()), {
+                            let mut error = e.to_string();
+                            for cause in e.chain() {
+                                error += &format!("\n  Caused by: {cause}");
+                            }
+                            error
```
Comment (charliermarsh) [inline crates/ruff_cli/src/commands/run.rs:92]: Does this manifest itself in practice? Like, what path / usage hits this?

File: crates/ruff/src/jupyter/schema.rs [hunk 27]
```diff
@@ -0,0 +1,256 @@
+//! The JSON schema of a Jupyter Notebook, entrypoint is [`JupyterNotebook`]
+//!
+//! Generated by <https://app.quicktype.io/> from
+//! <https://github.com/jupyter/nbformat/blob/16b53251aabf472ad9406ddb1f78b0421c014eeb/nbformat/v4/nbformat.v4.schema.json>
+//! Jupyter Notebook v4.5 JSON schema.
+//!
+//! Compared to the generated version, `Cell::id` is optional because it wasn't required <v4.5,
+//! I added `#[serde(deny_unknown_fields)]` where the schema had `"additionalProperties": false`
```
Comment (charliermarsh) [inline crates/ruff/src/jupyter/schema.rs:8]: For whatever reason I tend to use "we" instead of "I" in code comments, I don't know why :joy:

**Review by charliermarsh — APPROVED** [review]
🎸 🎸 🎸

File: crates/ruff/src/jupyter/schema.rs (hunk 27 above)
Comment (charliermarsh) [reply crates/ruff/src/jupyter/schema.rs:8]: I guess to me it conveys that we're all responsible for whatever is being written / committed

Comment (konstin) [reply crates/ruff/src/jupyter/schema.rs:8]: i normally use we for general explanations and i when i did something opionated, but i agree with the ~~royal~~ [mathematical](https://math.stackexchange.com/a/447722) we over i. here i think we can get away with passive voice.
